File: Algorithms/merge_sort.c
#include<stdio.h>
#include<conio.h>
//int arr[100];

void merge(int *arr,int beg,int end)
{
    int p=0,q=0,r=0,mid,tp,tq;
    int tmp1[10],tmp2[10];
    mid=(beg+end)/2;
    //printf("\n Mid: %d\n",mid);
    
    for(int i=beg;i<mid;i++)
    {
        tmp1[p++]=arr[i];
    }
    for(int j=mid;j<end;j++)
    {
        tmp2[q++]=arr[j];
    }
    tp=p-1;
    tq=q-1;
    p=0;
    q=0;
    r=beg;
    // printf("\n Left: ");
    // for(int i=beg;i<=mid;i++)
    // {
    //     printf("\t %d",tmp1[i]);
    // }
    // printf("\n Right: ");
    // for(int i=mid+1;i<=end;i++)
    // {
    //     printf("\t %d",tmp2[q++]); 
    // }

    while(p<=tp && q<=tq)
    {
        if(tmp1[p]<tmp2[q])
        {
            arr[r]=tmp1[p++];
        }
        else
        {
            arr[r]=tmp2[q++];
        }
        r++;
    }
    while(p<=tp)
    {
        arr[r++]=tmp1[p++];
    }
    while(q<=tq)
    {
        arr[r++]=tmp2[q++];
    }
    for(int i=0;i<5;i++)
    {
        printf("%d\t",arr[i]);
    }
    printf("\n");

}

void merge_sort(int *arr,int beg,int end)
{
    int i=0,j,cnt=0,ind,mid;
    mid = (beg+end)/2;
    if (beg<end)
    {
        printf("beg: %d , end: %d \n",beg,end);
        merge_sort(arr,beg,mid);
        merge_sort(arr,mid,end);
        merge(arr,beg,end);
    }
}

int main()
{
    int n,i,arr[10];
    printf("\n\t\tMerge Sort\t\t\n");
    printf("Enter the no. of elements : ");
    scanf("%d",&n);
    
    printf("Enter the %d elements : ",n);
    for(i=0;i<n;i++)
    {
        scanf("%d",&arr[i]);
    }
    merge_sort(arr,0,n);
    for(i=0;i<n;i++)
    {
        printf("%d\t",arr[i]);
    }
    getch();
    return 0;
}

File: Algorithms/merge_sort.py
def merge(l,beg,end):
    tmp1=[]
    tmp2=[]
    mid=(beg+end)//2
    p=0
    q=0
    r=beg
    for i in range(beg,mid+1):
        tmp1.append(l[i])
    for i in range(mid+1,end+1):
        tmp2.append(l[i])
    while(p<len(tmp1) and q<len(tmp2)):
        if tmp1[p]<tmp2[q]:
            l[r]=tmp1[p]
            p+=1
        else:
            l[r]=tmp2[q]
            q+=1
        r+=1
        
    while(p<len(tmp1)):
        l[r]=tmp1[p]
        p+=1
        r+=1
    while(q<len(tmp2)):
        l[r]=tmp2[q]
        q+=1
        r+=1
        
    

def merge_sort(l,beg,end):
    if beg<end:
        mid=(beg+end)//2
        merge_sort(l,beg,mid)
        merge_sort(l,mid+1,end)
        merge(l,beg,end)

if __name__ == "__main__":
    n=int(raw_input())
    l=list(map(int,raw_input().split()))
    merge_sort(l,0,n-1)
    print(l)
    